### PR TITLE
Enable to answer a request a copy using "helpdesk" strategy

### LINKED ISSUE #8304: Unable to answer a request a copy using "helpdesk" strategy in DSpace 7.2
**Describe the bug**
When request a copy is enabled with helpdesk settings, if we try to either approve or reject a request for a copy by anonymous user, an error 500 occurs (see log file for details: - [dspace.log](https://github.com/DSpace/DSpace/files/8763897/dspace.log)).

Our resquest a copy are configurations are as follow: 
request.item.type = all
mail.helpdesk = (valid email of our Helpdesk)
request.item.helpdesk.override = true

The error occurs even if the user have admin permissions on the collection to which the item requested belongs.

**To Reproduce**
1. Activate request a copy with Helpdesk Strategy: https://wiki.lyrasis.org/display/DSDOC7x/Request+a+Copy#RequestaCopy-(Optional)RequestingacopywithHelpDeskworkflow
2. As anonymous user, look at an embargoed item, click on the file to display the request a copy form.
3. Complete and send the request a copy form
4. As helpdesk staff, look at the email inbox
5. Open the Request a copy email and click on the link to respond to it.
6. Use your DSpace Credentials to connect (_First problem, as previously this was available anonymously_)
7. On the response form try to approve or decline the request.
8. An error message appears. (_Second problem, as it's not possible to submit the request when HelpDesk strategy is enabled_)

**Expected behavior**
No error messages, the database is updated to reflect if the request has been approved or rejected and the user received the email response and, if the request has been accepted, the item file(s).

**Related work**
Don't know if this problem could relate to #8199

### PR DESCRIPTION
**Pull request by mwoodiupui**
## References
* Fixes #8304

## Description
We no longer try to validate the current user for a copy request when using the "help desk" strategy.  Anyone who has the request email can use the link.  For security we rely on the discretion of the help desk personnel.

## Instructions for Reviewers
List of changes in this PR:
* Authorization for responding to a copy request is pushed down to the individual strategy classes
* The help desk strategy always returns `true` when asked for authorization.

You can test this with the procedure given in the linked Issue.

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by nibou230** [comment]
Hi @mwoodiupui, I looked and tested the code you provided.

The copy request can be executed properly with those changes and everything else seems good for us, great job!

Thank you very much!

**Review by abollini — CHANGES_REQUESTED** [review]
Hi @mwoodiupui thanks for this PR it looks really a great idea to me.
There is a slightly complication to consider due to the fact that in this other PR https://github.com/DSpace/DSpace/pull/2115 you propose to support potentially multiple strategy at once (the combining strategy). In a such scenario the combining strategy would be unable to proper delegate the check to real strategy that has decided to sent the token to the user this mean that if the help desk strategy is enabled anyone that has the email would be able to accept/reject it regardless to the authorization check in the other strategy.
As I have initially pointed in this comment (before to realize that this other PR works exactly on this area) https://github.com/DSpace/DSpace/pull/2115#discussion_r928146691 personally I'm also ok with a semplification where we simply trust anyone (including not logged-in people) that have a valid token. Otherwise, I guess that any strategy should generate a specific token bind to the strategy (so we should add a column in the database to store it or the strategy id should be part of the token) and separated email should be sent for each strategy. If this is the way to go we can probably remove the combining strategy and have spring to return all the implementation of the strategies moving the iteration directly in the RestRepository.

We need also to coordinate which PR comes in first, I would suggest to finalize the https://github.com/DSpace/DSpace/pull/2115 first

**Comment by mwoodiupui** [comment]
I agree.  I don't know why `isAuthorized` was created.  Request a Copy is meant to operate outside of DSpace's authorization scheme.

File: dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemSubmitterStrategy.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemSubmitterStrategy.java:27]: This logger doesn't appear to be used anywhere in this file. Maybe it should just be removed until it's needed?

File: dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemHelpdeskStrategy.java
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemHelpdeskStrategy.java:56]: Small thing, but you created `P_` variables above and then didn't use them in this class.  I see you use them in the tests, but maybe we should use them on lines 55 & 56 as well?

File: dspace-api/src/test/java/org/dspace/app/requestitem/RequestItemSubmitterStrategyTest.java
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/app/requestitem/RequestItemSubmitterStrategyTest.java:72]: Small thing, this line is duplicated.

**Review by tdonohue — COMMENTED** [review]
@mwoodiupui : Overall, this works well (much better than current "main")! But I ran into a _possible bug_ and have a few minor suggestions for improvements.

1. Possible Bug:  According to ticket #8304, when the "Help Desk" is enabled, it should be possible to respond to a copy request _without logging into DSpace_.  This doesn't work.  When I enable the help desk option, and click on the link sent to the `mail.helpdesk`, that link forces me to first login to DSpace.  After that, I can respond to the request.
    * If we feel this is more complex, I'm OK with us leaving this PR as-is, and creating a separate ticket to fix this in a later release. 
2. Possible improvements/updates to `requestitem.xml`.  I wonder if should [uncomment](https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/requestitem.xml#L47-L48) `RequestItemHelpdeskStrategy` by default now that you need to [update the implementation via the alias](https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/requestitem.xml#L23-L24)?   This seems like it'd make it easier to switch to the HelpDesk approach, though it would mean we also need to update the docs on enabling the Helpdesk approach here: https://wiki.lyrasis.org/display/DSDOC7x/Request+a+Copy#RequestaCopy-Configureallrequeststogotoahelpdeskemail

Beyond that, I have some minor notes inline on small things to possibly cleanup.  Nothing that impacts the feature though.

**Comment by tdonohue** [comment]
@mwoodiupui : Ah, ok.  I missed that discussion about commenting out the unused beans.  I'm OK then with keeping them commented out.  But, we definitely still need to update the docs on how to enable the "helpdesk strategy".  Currently they say uncommenting the bean is all you need to do.  But, obviously you also need to update the `alias` now. https://wiki.lyrasis.org/display/DSDOC7x/Request+a+Copy#RequestaCopy-Configureallrequeststogotoahelpdeskemail

**Comment by mwoodiupui** [comment]
@tdonohue Re:  login required:  Thank you for catching this oversight.  dspace-server-webapp was still protecting the endpoint although dspace-api no longer cares.  Protection removed.  Possession of the token should be the only barrier to approving/denying a copy request.

File: dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemHelpdeskStrategy.java
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemHelpdeskStrategy.java:56]: This code was removed in [f6ab8ba](https://github.com/DSpace/DSpace/pull/8319/commits/f6ab8ba1fbd42e0c1e77e1c2779c684eed0fc69a).

File: dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemSubmitterStrategy.java
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/app/requestitem/RequestItemSubmitterStrategy.java]: Thanks, already noted, now fixed.

File: dspace-api/src/test/java/org/dspace/app/requestitem/RequestItemSubmitterStrategyTest.java
Comment (mwoodiupui) [reply dspace-api/src/test/java/org/dspace/app/requestitem/RequestItemSubmitterStrategyTest.java]: Thanks for spotting that.  Fixed.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RequestItemRepositoryIT.java
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RequestItemRepositoryIT.java:536]: @mwoodiupui : Rather than just removing all the unauthenticated user tests, I'd recommend we add a single unauthenticated user test which *proves* that only the token is required.  

So, I'd expect that we'd change one of these to prove that a successful response is returned.  That way we have a test to prove that our `permitAll()` is working properly...and that test would fail if someone ever tries to change that `permitAll()` to something else.

Comment (mwoodiupui) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/RequestItemRepositoryIT.java:536]: Good idea.  I think this is done.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@mwoodiupui : I've retested this today.  While the code itself looks good (no obvious issues), I'm having difficulty getting it to work as expected.  I'm not sure if I'm doing something unexpected, or not.

1.  First, I've discovered I *MUST* add `autowire-candidate="true"` to this bean definition: https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/requestitem.xml#L47-L48  Otherwise, if I add the `RequestItemHelpdeskStrategy` [to the alias](https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/requestitem.xml#L24) then entire backend fails to boot up cause of spring bean errors
    * I suspect the same issue might occur if you try to use the `CollectionAdministratorsRequestItemStrategy` bean here (but I haven't had a chance to test it): https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/requestitem.xml#L53-L54 (as it is also missing an "autowire-candidate" attribute)
2.  Once I have the bean properly enabled, I can successfully send the request to my configured `mail.helpdesk` when I enable `request.item.helpdesk.override`.  However, if I click the link in the email, **I still am forced to authenticate.**   I'm not sure why that is, because your changes in this PR look correct...but, something is still obviously forcing authentication to occur (I haven't had a chance to debug it)

I think we definitely need to figure out a fix for 1 before merging this PR.  If we cannot figure out 2, then we could create a ticket for it & delay it for the next release.

**Comment by mwoodiupui** [comment]
`autowire-candidate` requirement is caused by our very tight `default-autowire-candidates` restrictions in some other configuration file.  Why do we do this?  What autowiring do we want to prevent?

I've reworked this quite a bit.  Since there's been no further comment on the compromise suggested above, I've implemented that:  all strategies are uncommented and lazy-loaded, so only the strategy configured for use will be instantiated (plus its configured collaborators in the case of the combining strategy) and you only have to edit the alias to select one.  The alias is autowired by name (@Resource) rather than class so that we don't get multiple matches while wiring.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@mwoodiupui : I tried testing this today, and with the recent changes, I cannot get the "helpdesk" strategy to work anymore.  Maybe I'm doing something wrong?  Here's what I tried:
1. First, in `requestitem.xml`, I set the `<alias>` to this
  ```
  <alias alias='requestItemAuthorExtractor'
           name='org.dspace.app.requestitem.RequestItemHelpdeskStrategy'/>
  ```
2. Then in my local.cfg, I set the following:
  ```
  mail.server.disabled = true
  request.item.helpdesk.override = true
  mail.helpdesk = helpdesk@example.com
  ```
3. Then, while watching the `dspace.log` (`tail -f dspace.log`), I submitted a Request for Copy.  But, it was logged as being sent to the **author email** instead of my configured `mail.helpdesk` email address.

Am I doing something wrong in configuring this?  Maybe I misunderstood your recent changes?

**Comment by mwoodiupui** [comment]
@tdonohue It may have loaded the wrong strategy.  I found a change that had not been pushed up, which specifically requests the alias name.  I've pushed that change.  I've tested with it in place, and with your configuration I get a copy-request email sent "To: helpdesk@example.com".

**Review by tdonohue — APPROVED** [review]
👍 @mwoodiupui : I retested this today and the HelpDesk Strategy works. **But, it still does require that I authenticate** when I click the response link (e.g. `http://localhost:4000/request-a-copy/[token]`).  I'm not sure what is triggering the requirement for authentication, since you have an IT which proves it should work.  But, I'm not able to get it to work on my end.

Nonetheless, I would not like to hold this PR up any longer, as it fixes the larger bug of this feature being completely broken.  So, I'm giving this a +1 as I think it's good enough to merge into 7.5. If we cannot figure out the forced authentication issues, we can always log that as a separate bug to be addressed at a later time.

**Comment by mwoodiupui** [comment]
@tdonohue I have not yet found anything which ought to be requiring login.  I think we should go ahead with this patch and have a separate issue for the login requirement.

I'm sorry this is late.  I had promised it for yesterday.

**Comment by tdonohue** [comment]
@mwoodiupui : No worries on the delayed response.  In that case, I'll merge this with +1 approval (as it works much better than the current code on `main`).  I've created a follow-up ticket at #8636 for the outstanding issue.

**Pull request merged**
